modules/ding_nodelist/ding_nodelist.module

```diff
@@ -0,0 +1,2477 @@
+  }
+
+  // Restore other form elements if this is not the first form display.
+  if (isset($form_state['restore'])) {
+    $form = array_merge($form_state['original_form'], $form);
+  }
+
+  // Remove some standard controls.
+  unset($form['override_title']);
+  unset($form['override_title_text']);
+  unset($form['override_title_markup']);
+
+  return $form;
+}
+
+/**
+ * Generate a form piece for dynamic node select form.
+ *
+ * @param string $key
+ *   Unique key for this piece.
+ * @param string $type
+ *   Node type.
+ * @param array $node
+ *   Array with the node oject.
+ * @param int $weight
+ *   Piece weight.
+ *
+ * @return array
+ *   Form structure.
+ */
+function _ding_nodelist_selected_node($key, $type, $node = array(), $weight = 0) {
+  $form = array(
+    '#tree' => TRUE,
+    '#weight' => $weight,
+  );
+
+  // Set default value for 'node'
+  $node_default = (!is_null($node) && isset($node['node'])) ? $node['node'] : '';
+  $form['node'] = array(
+    '#type' => 'textfield',
+    '#title' => t('Select new node to add to list'),
+    '#title_display' => 'invisible',
+    '#autocomplete_path' => 'ding_nodelist/autocomplete/' . $type,
+    '#parents' => array('selected_nodes', $key, 'node'),
+    '#default_value' => $node_default,
+  );
+
+  $form['nid'] = array(
+    '#type' => 'value',
+    '#value' => NULL,
+    '#parents' => array('selected_nodes', $key, 'nid'),
+  );
+
+  // Set default value for 'nid' if not empty field.
+  if (!is_null($node) && isset($node['nid']) && ctype_digit($node['nid'])) {
+    $form['nid']['#value'] = $node['nid'];
+  }
+  elseif (!is_null($node) && isset($node['node'])) {
+    $tmp_node = explode(':', $node['node']);
+    $tmp_node = array_pop($tmp_node);
+    $tmp_node = substr($tmp_node, 0, -1);
+    if (ctype_digit($tmp_node)) {
+      $node['nid'] = $tmp_node;
+    }
+  }
+
+  $form['type'] = array(
+    '#type' => 'value',
+    '#value' => $type,
+    '#parents' => array('selected_nodes', $key, 'type'),
+  );
+
+  $form['weight'] = array(
+    '#type' => 'weight',
+    '#title' => $node['nid'] !== '' ? t('Weight for node @label', array('@label' => $node['nid'])) : t('Weight for new choice'),
+    '#title_display' => 'invisible',
+    '#default_value' => $weight,
+    '#delta' => 50,
+    '#parents' => array('selected_nodes', $key, 'weight'),
+  );
+
+  $form['node_delete_' . $key] = array(
+    '#name' => 'node_delete_' . $key,
+    '#type' => 'submit',
+    '#title_display' => 'invisible',
+    '#value' => t('Delete'),
+    '#parents' => array('selected_nodes', $key, 'node_delete_' . $key),
+    '#submit' => array('ding_nodelist_node_delete_submit'),
+    '#limit_validation_errors' => array(),
+    '#ajax' => array(
+      'wrapper' => 'ding_nodelist-selected-nodes',
+      'callback' => 'ding_nodelist_js_add_node',
+      'method' => 'replace',
+      'effect' => 'fade',
+    ),
+  );
+  return $form;
+}
+
+/**
+ * Builds more link form.
```

Comment:
Remember to document arguments. Especially in situations where it is not a hook implementation.